### PR TITLE
Particle conserving u1 graph decomp

### PR DESCRIPTION
**Pull request by comp-phys-marc**
**Context:** We would like to port the `ParticleConservingU1` decomposition to the new graph based system.

**Description of the Change:** Adds a capture compatible graph based decomposition to the class. Adds corresponding tests.

**Benefits:** The template can be decomposed via the graph.

**Possible Drawbacks:** N/A

**Related ShortCut Stories:** [sc-103123]

### CODE REVIEW

File: pennylane/templates/layers/particle_conserving_u1.py [hunk 1]
```diff
@@ -351,4 +366,108 @@ def shape(n_layers, n_wires):
...
+):  # pylint: disable=arguments-differ
+    nm_wires = [wires[l : l + 2] for l in range(0, len(wires) - 1, 2)]
+    nm_wires += [wires[l : l + 2] for l in range(1, len(wires) - 1, 2)]
+    n_layers = math.shape(weights)[0]
+
+    if isinstance(wires, Wires):
+        wires = wires.labels
+        nm_wires = list(map(lambda w: w.labels if isinstance(w, Wires) else w, nm_wires))
+
+    if has_jax and capture.enabled():
```
Comment (astralcai) [inline pennylane/templates/layers/particle_conserving_u1.py]: Capture can't be enabled if jax isn't available right? The `has_jax` test seems redundant?

File: pennylane/templates/layers/particle_conserving_u1.py [hunk 2]
```diff
@@ -351,4 +366,108 @@ def shape(n_layers, n_wires):
...
+def _particle_conserving_u1_decomposition(
+    weights: list, wires: WiresLike, init_state: tuple[int]
+):  # pylint: disable=arguments-differ
+    nm_wires = [wires[l : l + 2] for l in range(0, len(wires) - 1, 2)]
+    nm_wires += [wires[l : l + 2] for l in range(1, len(wires) - 1, 2)]
+    n_layers = math.shape(weights)[0]
+
+    if isinstance(wires, Wires):
+        wires = wires.labels
+        nm_wires = list(map(lambda w: w.labels if isinstance(w, Wires) else w, nm_wires))
```
Comment (astralcai) [inline pennylane/templates/layers/particle_conserving_u1.py:432]: What does this do?

File: pennylane/templates/layers/particle_conserving_u1.py [hunk 3]
```diff
@@ -351,4 +366,108 @@ def shape(n_layers, n_wires):
...
+
+    if has_jax and capture.enabled():
+        nm_wires, weights, wires, init_state = (
+            jnp.array(nm_wires),
+            jnp.array(weights),
+            jnp.array(wires),
+            jnp.array(init_state),
```
Comment (astralcai) [inline pennylane/templates/layers/particle_conserving_u1.py]: `qml.math.array(..., like="jax")` should do the same thing without requiring you to import `jax` here.

File: pennylane/templates/layers/particle_conserving_u1.py (hunk 2 above)
Comment (comp-phys-marc) [reply pennylane/templates/layers/particle_conserving_u1.py:432]: Thanks for your question. If we are using capture, then it is going to be problematic to use `Wires`. Instead, we need to use the wire labels. Here we have a collection of `Wires`, and this line goes through each of them and turns them into their `labels`.

File: pennylane/templates/layers/particle_conserving_u1.py (hunk 1 above)
Comment (comp-phys-marc) [reply pennylane/templates/layers/particle_conserving_u1.py]: I can try just checking if capture is enabled then.

File: pennylane/templates/layers/particle_conserving_u1.py (hunk 3 above)
Comment (comp-phys-marc) [reply pennylane/templates/layers/particle_conserving_u1.py]: Cool!

File: pennylane/templates/layers/particle_conserving_u1.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/templates/layers/particle_conserving_u1.py:432]: Mind leaving a comment to that effect? I was about to ask the same thing.

Comment (comp-phys-marc) [reply pennylane/templates/layers/particle_conserving_u1.py:432]: Sure thing.

Comment (albi3ro) [reply pennylane/templates/layers/particle_conserving_u1.py:432]: So why is this needed here but not in other templates?

Comment (comp-phys-marc) [reply pennylane/templates/layers/particle_conserving_u1.py:432]: It is needed in other templates whenever they index into wires and I have implemented it many times by now.

Comment (comp-phys-marc) [reply pennylane/templates/layers/particle_conserving_u1.py:432]: Actually I've done it so many times I don't wonder if it is worth making some change to `Wires` itself.

File: pennylane/templates/layers/particle_conserving_u1.py [hunk 4]
```diff
@@ -351,4 +360,111 @@ def shape(n_layers, n_wires):
...
         return n_layers, n_wires - 1, 2
+
+
+def _particle_conserving_u1_resources(n_layers: int, num_wires: int):
+    # number of pairs of even-indexed of wires
+    num_nm_wires = (
+        math.floor(num_wires / 2) if num_wires % 2 == 0 else math.ceil((num_wires - 1) / 2)
+    )
+    # number of odd-indexed pairs of wires
+    num_nm_wires += math.floor((num_wires - 1) / 2)
```
Comment (dwierichs) [inline pennylane/templates/layers/particle_conserving_u1.py]: See `ParticleConservingU2` PR

```suggestion
    num_nm_wires = num_wires-1
```

**Review by dwierichs — APPROVED** [review]
Just one suggestion like for `ParticleConservingU2`.
I think we can do better in the decomposition of this template, but let's do this in a separate PR, if it is desired :)
